### PR TITLE
implemented neighboring agent discovery and agent to agent communication

### PR DESCRIPTION
**Pull request by poorva1209**
Documentation to test in the distributed sample app PR: https://github.com/GRIDAPPSD/gridappsd-sample-distributed-app/pull/6

### CODE REVIEW

File: gridappsd/field_interface/context.py [hunk 1]
```diff
@@ -1,36 +1,53 @@
...
-
-        response = gridappsd_obj.get_response(request_field_context_queue, request)
+        response = downstream_message_bus.get_response(t.context_request_queue(downstream_message_bus.id), request, timeout=10)
         return response
 
     @classmethod
-    def get_context_by_message_bus(cls, downstream_message_bus_id):
+    def get_context_by_message_bus(cls, downstream_message_bus: FieldMessageBus):
         """
         return agents/devices based on upstream and/or downstream message bus as input
```
Comment (craig8) [inline gridappsd/field_interface/context.py:20]: the comment indicates that you can use the upstream/or downstream bus, however the argument seems to suggest only downstream is available.

File: gridappsd/field_interface/gridappsd_field_bus.py [hunk 2]
```diff
@@ -36,11 +36,26 @@ def subscribe(self, topic, callback):
     def unsubscribe(self, topic):
         pass
 
-    def publish(self, data, topic: str = None):
+    def send(self, topic, message):
         """
         Publish device specific data to the concrete message bus.
```
Comment (craig8) [inline gridappsd/field_interface/gridappsd_field_bus.py:41]: topic: str, message: Any

File: gridappsd/field_interface/gridappsd_field_bus.py [hunk 3]
```diff
@@ -36,11 +36,26 @@ def subscribe(self, topic, callback):
...
         """
-        pass
+        if self.gridappsd_obj is not None:
+            self.gridappsd_obj.send(topic, message)
+            
+    def get_response(self, topic, message, timeout=5):
```
Comment (craig8) [inline gridappsd/field_interface/gridappsd_field_bus.py:46]: These two functions do the exact same thing.  Why should there be two of them?

File: gridappsd/field_interface/interfaces.py [hunk 4]
```diff
@@ -164,194 +165,25 @@ def unsubscribe(self, topic):
         pass
 
     @abstractmethod
-    def publish(self, data, topic: str = None):
+    def send(self, topic, data):
```
Comment (craig8) [inline gridappsd/field_interface/interfaces.py:168]: elsewhere we use topic message....which should it be?

File: gridappsd/field_interface/interfaces.py [hunk 5]
```diff
@@ -164,194 +165,25 @@ def unsubscribe(self, topic):
...
-        pass
-
-    @abstractmethod
-    def unsubscribe(self, topic):
-        pass
-
+        
     @abstractmethod
-    def publish(self, data, topic: str = None):
+    def get_agent_response(self, agent_id, message, timeout):
```
Comment (craig8) [inline gridappsd/field_interface/interfaces.py:181]: abstractmethods don't have implementation....

**Review by craig8 — CHANGES_REQUESTED** [review]

File: gridappsd/field_interface/__init__.py [hunk 6]
```diff
@@ -1,9 +1,7 @@
 from typing import List
 
-from gridappsd.field_interface.context import ContextManager
```
[the remark this answers is not in the document]
Comment (poorva1209) [reply gridappsd/field_interface/__init__.py:3]: Updated to include both LocalContext and MessageBusDefinition

File: gridappsd/field_interface/agents/agents.py [hunk 7]
```diff
@@ -52,8 +46,16 @@ def __init__(self,
...
         self.simulation_id = simulation_id
         self.context = None
+        
+        #TODO: Change params and connection to local connection 
         self.params = ConnectionParameters()
         self.connection = GridappsdConnection(self.params)
+        
+        self.app_id = agent_config['app_id']
+        self.description = agent_config['description']
+        self.agent_id = str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (poorva1209) [reply gridappsd/field_interface/agents/agents.py]: Changed to 'da_[app_id]_[timestamp]'. Similar to instance id for centralized apps and services.

File: gridappsd/field_interface/agents/agents.py [hunk 8]
```diff
@@ -73,57 +75,95 @@ def __init__(self,
...
+        self.subscribe_to_messages()
+        self.subscribe_to_requests()
+        
+        if('context_manager' not in self.app_id):
+            LocalContext.register_agent(self.downstream_message_bus, self.upstream_message_bus,self)
 
     def subscribe_to_measurement(self):
         if self.simulation_id is None:
-            self.downstream_message_bus.subscribe(f"fieldbus/{self.downstream_message_bus.id}", self.on_measurement)
+            self.downstream_message_bus.subscribe(f"/topic/goss.gridappsd.field.output.{self.downstream_message_bus.id}", self.on_measurement)
```
[the remark this answers is not in the document]
Comment (poorva1209) [reply gridappsd/field_interface/agents/agents.py]: Updated to be imported from topics.

File: gridappsd/field_interface/agents/agents.py [hunk 9]
```diff
@@ -73,57 +75,95 @@ def __init__(self,
...
+    def on_request_from_uptream(self, headers: Dict, message):
+        self.on_request(self.upstream_message_bus, headers, message)
+    
+    def on_request_from_downstream(self, headers: Dict, message):
+        self.on_request(self.downstream_message_bus, headers, message)
+    
+    def on_request(self, message_bus, headers:Dict, message):
+        raise NotImplementedError(f"{self.__class__.__name__} must be overriden in child class")
+    
+    def get_registration_details(self):
```
[the remark this answers is not in the document]
Comment (poorva1209) [reply gridappsd/field_interface/agents/agents.py:165]: AgentRegitrationDetails dataclass added.

File: gridappsd/field_interface/context.py (hunk 1 above)
Comment (poorva1209) [reply gridappsd/field_interface/context.py]: Comment changed.

File: gridappsd/field_interface/context.py [hunk 10]
```diff
@@ -1,36 +1,53 @@
...
-        request = {'downstream_message_bus_id': downstream_message_bus_id,
+        
+        request = {'request_type' : 'get_context',
+                   'downstream_message_bus_id': downstream_message_bus.id,
                    'agents': True,
                    'devices': True}
+        return downstream_message_bus.get_response(t.context_request_queue(downstream_message_bus.id), request)
+    
+    @classmethod
+    def register_agent(cls, downstream_message_bus: FieldMessageBus, upstream_message_bus: FieldMessageBus, agent):
```
[the remark this answers is not in the document]
Comment (poorva1209) [reply gridappsd/field_interface/context.py:32]: That leads to circular dependency so will have to change in another way.

File: gridappsd/field_interface/gridappsd_field_bus.py (hunk 2 above)
Comment (poorva1209) [reply gridappsd/field_interface/gridappsd_field_bus.py:42]: Done.

File: gridappsd/field_interface/gridappsd_field_bus.py (hunk 3 above)
Comment (poorva1209) [reply gridappsd/field_interface/gridappsd_field_bus.py:47]: Removed duplicate. Will be good to have a get_agent_response but will add later.

File: gridappsd/field_interface/interfaces.py (hunk 4 above)
Comment (poorva1209) [reply gridappsd/field_interface/interfaces.py]: Changed to message.

File: gridappsd/field_interface/interfaces.py (hunk 5 above)
Comment (poorva1209) [reply gridappsd/field_interface/interfaces.py:180]: Removed abstract.

File: gridappsd/topics.py [hunk 11]
```diff
@@ -171,3 +174,57 @@ def simulation_log_topic(simulation_id):
...
+def field_message_bus_topic(message_bus_id, app_id=None, agent_id=None):
+    """ Utility method for getting the publish/subscribe topic for a specific message bus.
+
+    :param message_bus_id:
+    :param app_id:
+    :param agent_id:
+    :return:
+    """
+    assert message_bus_id, "message_bus_id cannot be empty"
+    return "{}.{}.{}.{}".format(BASE_FIELD_TOPIC, message_bus_id, app_id, agent_id)
```
[the remark this answers is not in the document]
Comment (poorva1209) [reply gridappsd/topics.py]: Changed. We do the same in all other functions so may have to change all . Will create another issue if needed.

**Pull request merged**